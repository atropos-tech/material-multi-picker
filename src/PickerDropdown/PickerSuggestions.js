import React from "react";
import { MenuItem, withStyles } from "@material-ui/core";
import { LOADING, NOT_ENOUGH_CHARACTERS, isError, suggestionsPropType } from "../utils";
import { string, func, number, any, object } from "prop-types";
import DefaultError from "./DefaultError";
import DefaultSuggestion from "./DefaultSuggestion";
import DefaultEmptyMessage from "./DefaultEmptyMessage";
import DefaultLoadingMessage from "./DefaultLoadingMessage";
import DefaultMoreCharactersMessage from "./DefaultMoreCharactersMessage";

const HIGHLIGHT_GREY_SHADE = 300;

const styles = theme => ({
    highlighted: {
        backgroundColor: theme.palette.grey[HIGHLIGHT_GREY_SHADE]
    }
});

function PickerSuggestions(
    { suggestions, getItemProps, highlightedIndex, itemToString, inputValue, SuggestionComponent = DefaultSuggestion, ErrorComponent = DefaultError, classes }
) {
    if ( isError(suggestions) ) {
        return (<ErrorComponent error={ suggestions } inputValue={ inputValue } />);
    }
    if ( suggestions === NOT_ENOUGH_CHARACTERS ) {
        return <DefaultMoreCharactersMessage />;
    }
    if ( suggestions === LOADING ) {
        return <DefaultLoadingMessage inputValue={ inputValue } />;
    }
    if ( Array.isArray(suggestions) ) {
        if ( suggestions.length ) {
            return (
                <>
                    {
                        suggestions.map((item, index) => {
                            const itemId = itemToString(item);
                            const isHighlighted = highlightedIndex === index;

                            const menuItemProps = getItemProps({
                                index,
                                item,
                                role: "menuitem",
                                className: isHighlighted ? classes.highlighted : "",
                                style: {
                                    padding: 0,
                                    height: "auto"
                                },
                            });
                            return (
                                <MenuItem className="suggestion" key={ itemId } {...menuItemProps}>
                                    <SuggestionComponent itemId={ itemId } item={ item } isHighlighted={ isHighlighted } inputValue={ inputValue } />
                                </MenuItem>
                            );
                        })
                    }
                </>
            );
        }
        if ( inputValue.length ) {
            return <DefaultEmptyMessage inputValue={ inputValue } />;
        } else {
            return false;
        }
    }

    console.error(suggestions);
    throw new Error("should never happen!");
}

PickerSuggestions.propTypes = {
    suggestions: suggestionsPropType,
    getItemProps: func.isRequired,
    highlightedIndex: number,
    itemToString: func.isRequired,
    inputValue: string.isRequired,
    SuggestionComponent: any,
    ErrorComponent: any,
    classes: object
};

export default withStyles(styles)(PickerSuggestions);
