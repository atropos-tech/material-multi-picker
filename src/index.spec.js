import React from "react";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { resetIdCounter } from "downshift";
import MultiPicker from "./index";
import { Chip, MenuItem, SvgIcon, TextField } from "@material-ui/core";
import JssProvider from "react-jss/lib/JssProvider";
import { BACKSPACE_KEYCODE } from "./utils";
import PickerSuggestions from "./PickerDropdown/PickerSuggestions";

// workaround for non-stable classnames generated in JSS
// https://github.com/mui-org/material-ui/issues/9492#issuecomment-368205258
//
// forces deterministic classnames that may clash in a wider app (but are safe in unit tests)
//
// Should be addressed in future material-ui release, see
//   https://github.com/mui-org/material-ui/issues/14357
function mountStable(reactElement, ...other) {
    const generateClassName = (rule, styleSheet) => `${styleSheet.options.classNamePrefix}-${rule.key}`;
    const elementWithStableClassNames = (
        <JssProvider generateClassName={ generateClassName }>
            { reactElement }
        </JssProvider>
    );
    return mount(elementWithStableClassNames, ...other);
}

const NOOP = () => { /* do nothing */ };
const SHORT_DELAY_MILLISECONDS = 30;

function delay(delayInMilliseconds = SHORT_DELAY_MILLISECONDS) {
    return new Promise(resolve => setTimeout(resolve, delayInMilliseconds) );
}

const BASE_PROPS = {
    value: [],
    onChange: NOOP,
    itemToString: item => item,
    getSuggestedItems: () => [],
    disablePortals: true
};

async function changeInputValueAndUpdate(wrapper, newInputValue) {
    act(() => {
        wrapper.find("input").simulate("change", { target: { value: newInputValue}});
    });
    await delay(200);
    wrapper.update();
}

describe("MultiPicker component", () => {

    // avoid non-deterministic id creation for downshift subcomponents
    // breaking snapshot testing
    beforeEach(resetIdCounter);

    it("renders empty content", () => {
        expect.assertions(1);

        const wrapper = mountStable(<MultiPicker { ...BASE_PROPS } />);
        expect(wrapper.getDOMNode()).toMatchSnapshot();
    });

    it("renders single chip", () => {
        expect.assertions(3);

        const wrapper = mountStable(<MultiPicker { ...BASE_PROPS } value={ ["some item"] } />);
        expect(wrapper).toContainExactlyOneMatchingElement(Chip);
        expect(wrapper.find(Chip)).toHaveText("some item");
        expect(wrapper.find(TextField).getDOMNode()).toMatchSnapshot();
    });

    it("can be disabled", () => {
        expect.assertions(2);
        const wrapper = mountStable(<MultiPicker { ...BASE_PROPS } disabled value={["some item"]} />);
        expect(wrapper.find("input")).toBeDisabled();
        expect(wrapper.find(Chip)).not.toContainMatchingElement(SvgIcon);
    });

    it("can show in error state", async () => {
        expect.assertions(2);
        const wrapper = mountStable(<MultiPicker { ...BASE_PROPS } error value={["some item"]} />);
        await delay(50);
        expect(wrapper.find(TextField).getDOMNode()).toMatchSnapshot();
        expect(wrapper.find(TextField)).toHaveProp("error", true);
    });

    it("clears the input on blur if the clearInputOnBlur prop is set", async () => {
        expect.assertions(1);
        const wrapper = mountStable(<MultiPicker { ...BASE_PROPS } clearInputOnBlur />);

        await changeInputValueAndUpdate(wrapper, "some text");
        wrapper.find("input").simulate("blur");
        wrapper.update();

        expect(wrapper.find("input")).toHaveProp("value", "");
    });

    it("does not clear the input on blur if the clearInputOnBlur prop is set", async () => {
        expect.assertions(1);
        const wrapper = mountStable(<MultiPicker {...BASE_PROPS }/>);

        await changeInputValueAndUpdate(wrapper, "some text");
        wrapper.find("input").simulate("blur");
        wrapper.update();

        expect(wrapper.find("input")).toHaveProp("value", "some text");
    });

    it("renders dropdown when typing", async () => {
        expect.assertions(4);

        const wrapper = mountStable(<MultiPicker { ...BASE_PROPS } getSuggestedItems={ () => ["some suggestion"] } />);
        expect(wrapper).not.toContainMatchingElement(PickerSuggestions);

        await changeInputValueAndUpdate(wrapper, "some text");

        expect(wrapper).toContainExactlyOneMatchingElement(PickerSuggestions);
        expect(wrapper.find(PickerSuggestions)).toHaveText("some suggestion");
        expect(wrapper.find(PickerSuggestions).getDOMNode()).toMatchSnapshot();
    });

    it("adds the correct item when it is clicked", async () => {
        expect.assertions(2);

        const onChange = jest.fn();
        const getSuggestedItems = () => ["some suggestion", "some other suggestion"];
        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } onChange={ onChange } value={ ["some item"] } getSuggestedItems={ getSuggestedItems } />);
        await changeInputValueAndUpdate(wrapper, "some text");

        expect(wrapper).toContainMatchingElements(2, MenuItem);

        wrapper.find(MenuItem).at(1).simulate("click");
        expect(onChange).toHaveBeenCalledWith(["some item", "some other suggestion"]);
    });

    it("removes items when the chip remove icon is clicked", () => {
        expect.assertions(2);

        const onChange = jest.fn();
        const getSuggestedItems = () => ["some suggestion", "some other suggestion"];
        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } onChange={ onChange } value={ ["some item", "some-other-item"] } getSuggestedItems={ getSuggestedItems } />);

        expect(wrapper).toContainMatchingElements(2, Chip);

        // simulating click doesn't work because Chip uses SvgIcon and you
        // can't simulate clicks on <svg> elements!
        act(() => {
            wrapper.find(Chip).at(0).props().onDelete();
        });

        expect(onChange).toHaveBeenCalledWith(["some-other-item"]);
    });

    it("removes last item on backspace", () => {
        expect.assertions(1);

        const onChange = jest.fn();
        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } onChange={ onChange } value={ ["some item", "some-other-item"] } />);
        wrapper.find("input").simulate("keydown", { keyCode: BACKSPACE_KEYCODE });

        expect(onChange).toHaveBeenCalledWith(["some item"]);
    });

    it("does nothing on backspace if there are no items", () => {
        expect.assertions(1);

        const props = {
            ...BASE_PROPS,
            onChange: jest.fn(NOOP),
            getSuggestedItems: () => []
        };
        const wrapper = mountStable(<MultiPicker {...props }/>);
        wrapper.find("input").simulate("keydown", { keyCode: BACKSPACE_KEYCODE });

        expect(props.onChange).not.toHaveBeenCalled();
    });

    it("triggers the 'onFocus' callback when the event occurs", () => {
        expect.assertions(1);

        const onFocus = jest.fn();
        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } onFocus={ onFocus } />);
        wrapper.find("input").simulate("focus");

        expect(onFocus).toHaveBeenCalled();
    });

    it("triggers the 'onBlur' callback when the event occurs", () => {
        expect.assertions(1);

        const onBlur = jest.fn();
        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } onBlur={ onBlur } />);
        wrapper.find("input").simulate("blur");

        expect(onBlur).toHaveBeenCalled();
    });

    it("triggers the 'onDragStart' callback when the event occurs", () => {
        expect.assertions(1);

        const onDragStart = jest.fn();
        const wrapper = mountStable(<MultiPicker {...BASE_PROPS } onDragStart={ onDragStart } />);
        wrapper.find("input").simulate("dragstart");

        expect(onDragStart).toHaveBeenCalled();
    });
});
