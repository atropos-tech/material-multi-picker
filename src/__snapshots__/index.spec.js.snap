// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiPicker component can show in error state 1`] = `
<div
  class="MuiFormControl-root"
>
  
  <div
    class="MuiInputBase-root MuiInput-root PickerInput-InputRoot MuiInput-underline MuiInputBase-error MuiInput-error MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
  >
    <div
      class="MuiChip-root MuiChip-deletable MultiPicker-pickerChip"
      role="button"
      tabindex="-1"
    >
      <span
        class="MuiChip-label"
      >
        some item
      </span>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiChip-deleteIcon"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
        />
      </svg>
    </div>
    <input
      aria-autocomplete="list"
      aria-invalid="true"
      aria-labelledby="downshift-0-label"
      autocomplete="off"
      class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart PickerInput-inputRoot"
      id="downshift-0-input"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`MultiPicker component renders dropdown when typing 1`] = `
<li
  aria-selected="false"
  class="MuiButtonBase-root MuiListItem-root MuiListItem-default MuiListItem-gutters MuiListItem-button MuiMenuItem-root MuiMenuItem-gutters"
  id="downshift-0-item-0"
  role="menuitem"
  style="padding: 0px; height: auto;"
  tabindex="-1"
>
  <p
    class="MuiTypography-root MuiTypography-body1"
    style="padding: 11px 16px;"
  >
    some suggestion
  </p>
  <span
    class="MuiTouchRipple-root"
  />
</li>
`;

exports[`MultiPicker component renders empty content 1`] = `
<div
  aria-expanded="false"
  aria-haspopup="listbox"
  aria-labelledby="downshift-0-label"
  role="combobox"
  style="position: relative;"
>
  <div
    class="MuiFormControl-root"
  >
    
    <div
      class="MuiInputBase-root MuiInput-root PickerInput-InputRoot MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-autocomplete="list"
        aria-invalid="false"
        aria-labelledby="downshift-0-label"
        autocomplete="off"
        class="MuiInputBase-input MuiInput-input PickerInput-inputRoot"
        id="downshift-0-input"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`MultiPicker component renders single chip 1`] = `
<div
  class="MuiFormControl-root"
>
  
  <div
    class="MuiInputBase-root MuiInput-root PickerInput-InputRoot MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
  >
    <div
      class="MuiChip-root MuiChip-deletable MultiPicker-pickerChip"
      role="button"
      tabindex="-1"
    >
      <span
        class="MuiChip-label"
      >
        some item
      </span>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiChip-deleteIcon"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
        />
      </svg>
    </div>
    <input
      aria-autocomplete="list"
      aria-invalid="false"
      aria-labelledby="downshift-0-label"
      autocomplete="off"
      class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart PickerInput-inputRoot"
      id="downshift-0-input"
      type="text"
      value=""
    />
  </div>
</div>
`;
